### PR TITLE
Refactor `Authentication` methods based on the latest API reference

### PR DESCRIPTION
**Pull request by LauraBeatris**
## Description

- Updates all of the authentication methods to follow the following nomenclature: `authenticates_with_${method}`
- Adds `authenticate_with_email_verification`

## Documentation

Does this require changes to the WorkOS Docs? E.g. the [API Reference](https://workos.com/docs/reference) or code snippets need updates.

```
[X] Yes
```

If yes, link a related docs PR and add a docs maintainer as a reviewer. Their approval is required.

### CODE REVIEW

**Comment by LauraBeatris** [comment]
Couple of things to check on Monday:

- `authentication_challenge_id` is not mentioned in the API reference for CURL, but added to other SDK methods, and I also see this param on our internal API DTO 
- `invite_token` on the authorization code, password and magic code DTOs, however it's not mentioned on our API reference 
- `link_authorization_code` is also defined in the magic code internal DTO, but not in the API reference, I'd presume in this case that it's related to AuthKit only, right?

**Comment by LauraBeatris** [comment]
Struggling to perform QA with `authenticate_with_code` due to some SSO issues, going to come back to this tomorrow once I discuss with the team.

**Comment by LauraBeatris** [comment]
Also to note, we might need to fix the docs reference for `authenticate_with_magic_auth` cause it's showing password authentication atm:  

CleanShot 2023-11-19 at 17 24 24@2x

**Comment by LauraBeatris** [comment]
The Ruby snippet for `authenticate_with_magic_auth` in the docs, isn't including the `user_id` param:

CleanShot 2023-11-19 at 19 09 21@2x

File: lib/workos/user_management.rb [hunk 1]
```diff
@@ -481,19 +489,30 @@ def authenticate_user_with_code(code:, client_id:, ip_address: nil, user_agent:
           client_id: String,
           pending_authentication_token: String,
           authentication_challenge_id: String,
+          ip_address: T.nilable(String),
+          user_agent: T.nilable(String),
         ).returns(WorkOS::UserResponse)
       end
-      def authenticate_user_with_totp(code:, client_id:, pending_authentication_token:, authentication_challenge_id:)
+      def authenticate_with_totp(
```
Comment (LauraBeatris) [inline lib/workos/user_management.rb:496]: I believe this is still in ongoing development? CC @amygdalama

Comment (amygdalama) [reply lib/workos/user_management.rb:496]: yeah i assume we are shipping with this but it was missing from the docs. i believe dane was gonna ask hadley about it today.

File: lib/workos/user_management.rb [hunk 2]
```diff
@@ -481,19 +489,30 @@ def authenticate_user_with_code(code:, client_id:, ip_address: nil, user_agent:
...
         response = execute_request(
           request: post_request(
-            path: '/users/authenticate',
+            path: '/user_management/authenticate',
             body: {
               code: code,
               client_id: client_id,
               client_secret: WorkOS.config.key!,
               pending_authentication_token: pending_authentication_token,
+              grant_type: 'urn:workos:oauth:grant-type:email-verification:code',
```
Comment (amygdalama) [inline lib/workos/user_management.rb:512]: was this change intentional?

Comment (LauraBeatris) [reply lib/workos/user_management.rb:512]: Not sure if it's right? I followed the API reference: 

CleanShot 2023-11-20 at 18 40 42

CleanShot 2023-11-20 at 18 40 37

But it's confusing, it seems that we're missing a section there.

Comment (amygdalama) [reply lib/workos/user_management.rb:512]: yup there's a bug in the docs; `urn:workos:oauth:grant-type:mfa-totp` is the correct grant type

**Review by amygdalama — APPROVED** [review]
once the grant type is fixed this is good to merge! thank you 🙌
